app.go

```diff
@@ -113,6 +113,40 @@ func (o provideOption) String() string {
 	return fmt.Sprintf("fx.Provide(%s)", strings.Join(items, ", "))
 }
 
+// Supply makes the given values available for dependency resolution as if
+// they had been provided by a constructor. Specifically, given:
+//
+//  type MyStruct struct{}
+//  var myStruct = &MyStruct{}
+//
+// The following two forms are equivalent:
+//
+//  fx.Provide(func() *MyStruct { return myStruct })
+//  fx.Supply(myStruct)
+//
+// Supply operates by constructing a constructor of the first form based on
+// the types of the supplied values. Supply accepts any number of arguments.
+func Supply(values ...interface{}) Option {
+	if len(values) == 0 {
+		return Options()
+	}
+
+	returnTypes := make([]reflect.Type, len(values))
+	returnValues := make([]reflect.Value, len(values))
+
+	for i, value := range values {
+		returnTypes[i] = reflect.TypeOf(value)
```

Comment:
Some validation:

- If `reflect.ValueOf(value).IsZero()`, they passed in nil. We should panic (because bad usage).
- Since `fx.Provide` treats `error` returns differently, we should verify here that the type is not an `error` type. We can use [`fx.Error`][1] to short-circuit this and fail `fx.New`.
  
  [1]: https://pkg.go.dev/go.uber.org/fx?tab=doc#Error

Reply:
We'll also need to support [fx.Annotated][1] since that will be the only way
users will be able to use named values or value groups with fx.Supply.

[1]: https://pkg.go.dev/go.uber.org/fx?tab=doc#Annotated

```
fx.Supply(fx.Annotated{Target: v, Name: "foo})
```

To support it, we can do something like this:

```
for _, v := range values {
  var c interface{}
  if ann, ok := v.(Annotated); ok {
    ann.Target = constructorForValue(ann.Target)
    c = ann
  } else {
    c = constructorForValue(c)
  }
  ..
```

Where `constructorForValue` builds the `func() $Type` constructor.

This has one minor change. Instead of, `fx.Supply(1, "foo", true)` being
equivalent to,

    fx.Provide(func() (int, string, bool) {
      return 1, "foo", true
    })

It will now be equivalent to,

    fx.Provide(
      func() int    { return 1     },
      func() string { return "foo" },
      func() bool   { return true  },
    )

That way, `fx.Supply("foo", fx.Annotated{Target: "bar", Name: "x"})` becomes,

    fx.Provide(
      func() string { return "foo" },
      fx.Annotated{
        Target: func() string { return "foo" },
        Name: "x",
      },
    )

Reply:
@abhinav I've introduced the proposed changes, but with the following differences:

- Panicking specifically on `interface{}(nil)`, not zero values in general, as zero values (such as empty structs) could be supplied for valid reasons.
- Panicking on error values rather than failing the app container, as supplied values are constructed (and thus validated) outside of a `fx.App`'s runtime. If a user had an error on hand when calling `fx.Supply`, they ought to have checked it earlier. This one's a semantic decision though, so I'm happy to change to `fx.Error` if that's more in line with what users would expect.

Reply:
> Panicking specifically on `interface{}(nil)`, not zero values in general, as zero values (such as empty structs) could be supplied for valid reasons.

Oops, I meant `Value.IsValid` (which would be true for `interface{}(nil)`).

RE: the second point: That's reasonable. There's no valid use case for `fx.Supply(err)` where `err` is known to be an `error`.